sdk/python/lib/pulumi_policy/policy.py

```diff
@@ -0,0 +1,751 @@
+        sys.stdout.buffer.write(f"{port}\n".encode())
+        try:
+            while True:
+                time.sleep(_ONE_DAY_IN_SECONDS)
+        except KeyboardInterrupt:
+            server.stop(0)
+
+
+class EnforcementLevel(Enum):
+    """
+    Indicates the impact of a policy violation.
+    """
+
+    ADVISORY = "advisory"
+    MANDATORY = "mandatory"
+    DISABLED = "disabled"
+
+
+class Policy(ABC):
+    """
+    A policy function that returns true if a resource definition violates some policy (e.g., "no
+    public S3 buckets"), and a set of metadata useful for generating helpful messages when the policy
+    is violated.
+    """
+
+    name: str
+    """
+    An ID for the policy. Must be unique within the current policy set.
+    """
+
+    description: str
+    """
+    A brief description of the policy rule. e.g., "S3 buckets should have default encryption
+    enabled."
+    """
+
+    enforcement_level: Optional[EnforcementLevel]
+    """
+    Indicates what to do on policy violation, e.g., block deployment but allow override with
+    proper permissions.
+    """
+
+    def __init__(self,
+                 name: str,
+                 description: str,
+                 enforcement_level: Optional[EnforcementLevel] = None) -> None:
+        """
+        :param str name: An ID for the policy. Must be unique within the current policy set.
+        :param str description: A brief description of the policy rule. e.g., "S3 buckets should have
+               default encryptionenabled."
+        :param Optional[EnforcementLevel] enforcement_level: Indicates what to do on policy violation,
+               e.g., block deployment but allow override with proper permissions.
+        """
+        if not name:
+            raise TypeError("Missing name argument")
+        if not isinstance(name, str):
+            raise TypeError("Expected name to be a string")
+        if name == "all":
+            raise TypeError(
+                'Invalid policy name "all"; "all" is a reserved name')
+        if not description:
+            raise TypeError("Missing description argument")
+        if not isinstance(description, str):
+            raise TypeError("Expected description to be a string")
+        if enforcement_level is not None and not isinstance(enforcement_level, EnforcementLevel):
+            raise TypeError(
+                "Expected enforcement_level to be an EnforcementLevel")
+        self.name = name
+        self.description = description
+        self.enforcement_level = enforcement_level
+
+
+ReportViolation = Callable[[str, Optional[str]], None]
+"""
+ReportViolation is the callback signature used to report policy violations.
+The first param is the violation message and the second is an optional
+urn of the resource to associate with the violations.
+"""
+
+
+class ResourceValidationArgs:
+    """
+    ResourceValidationArgs is the argument bag passed to a resource validation.
+    """
+
+    resource_type: str
```

Comment:
Ideally, this would be `type`, but `type` is reserved, so we could either go with `type_` or something more meaningful like `resource_type`. Luke and I chatted about this and decided to go with the latter, even though that's different than what we've done with resource transformations [here](https://github.com/pulumi/pulumi/blob/cfcfad2f3c8c4bb7f34130f68e2c842dda3dc7e6/sdk/python/lib/pulumi/resource.py#L222). (We probably should have used `resource_type` for transformations.)